File: 2022/Day11/main.py
with open('input.txt', 'r') as f:
    data = f.read()


from copy import copy
from itertools import accumulate

test = """Monkey 0:
  Starting items: 79, 98
  Operation: new = old * 19
  Test: divisible by 23
    If true: throw to monkey 2
    If false: throw to monkey 3

Monkey 1:
  Starting items: 54, 65, 75, 74
  Operation: new = old + 6
  Test: divisible by 19
    If true: throw to monkey 2
    If false: throw to monkey 0

Monkey 2:
  Starting items: 79, 60, 97
  Operation: new = old * old
  Test: divisible by 13
    If true: throw to monkey 1
    If false: throw to monkey 3

Monkey 3:
  Starting items: 74
  Operation: new = old + 3
  Test: divisible by 17
    If true: throw to monkey 0
    If false: throw to monkey 1"""


class Monkey:
    def __init__(self, items, op, test, test_true, test_false):
        self.items = items
        self.op = op
        self.test = test
        self.tt = test_true
        self.tf = test_false
        self.inspected = 0

    def execute(self, other_monkeys):
        temp = copy(self.items)
        for old in temp:
            new = eval(self.op)
            new = new // 3
            rem = new % self.test
            if rem == 0:
                other_monkeys[self.tt].items.append(new)
            else:
                other_monkeys[self.tf].items.append(new)

            self.inspected += 1

            self.items.pop(0)

    def __repr__(self):
        return f'Items: {",".join(list(map(str, self.items))):<25} Operation: {self.op:<10} Test: {self.test:<5} Test True: {self.tt:<5} Test False: {self.tf:<5} Inspected: {self.inspected:<5}'


def get_part_one(data):
    monkeys = []
    monkeys_raw = data.split('\n\n')

    for monkey_raw in monkeys_raw:
        lines = monkey_raw.split('\n')
        items = list(map(int, lines[1].split(':')[1].split(',')))
        op = lines[2].split('=')[1].strip()
        t = int(lines[3].split(' ')[-1])
        tt = int(lines[4].split(' ')[-1])
        tf = int(lines[5].split(' ')[-1])

        mon = Monkey(items, op, t, tt, tf)
        monkeys.append(mon)

    for r in range(20):
        for mon in monkeys:
            mon.execute(monkeys)

    act = []
    for mon in monkeys:
        act.append(mon.inspected)

    act.sort()

    act = act[-2:]

    prod = 1

    for a in act:
        prod *= a

    return prod


assert get_part_one(test) == 10605
print(f'Part 1: {get_part_one(data)}')

# part 2


class Monkey2:
    def __init__(self, items, op, test, test_true, test_false):
        self.items = items
        self.op = op
        self.test = test
        self.tt = test_true
        self.tf = test_false
        self.inspected = 0
        self.lcm = None

    def execute(self, other_monkeys):
        temp = copy(self.items)
        for old in temp:
            new = eval(self.op)
            new = new % self.lcm
            rem = new % self.test
            if rem == 0:
                other_monkeys[self.tt].items.append(new)
            else:
                other_monkeys[self.tf].items.append(new)

            self.inspected += 1

            self.items.pop(0)

    def __repr__(self):
        return f'Items: {",".join(list(map(str, self.items))):<25} Operation: {self.op:<10} Test: {self.test:<5} Test True: {self.tt:<5} Test False: {self.tf:<5} Inspected: {self.inspected:<5}'


def get_part_two(data):
    monkeys = []
    monkeys_raw = data.split('\n\n')

    lcm = 1

    for monkey_raw in monkeys_raw:
        lines = monkey_raw.split('\n')
        items = list(map(int, lines[1].split(':')[1].split(',')))
        op = lines[2].split('=')[1].strip()
        t = int(lines[3].split(' ')[-1])
        tt = int(lines[4].split(' ')[-1])
        tf = int(lines[5].split(' ')[-1])
        lcm *= t

        mon = Monkey2(items, op, t, tt, tf)
        monkeys.append(mon)

    for mon in monkeys:
        mon.lcm = lcm

    for r in range(10000):
        for mon in monkeys:
            mon.execute(monkeys)

    act = []
    for mon in monkeys:
        act.append(mon.inspected)

    act.sort()

    act = act[-2:]

    prod = 1

    for a in act:
        prod *= a

    return prod


assert get_part_two(test) == 2713310158
print(f'Part 2: {get_part_two(data)}')

File: 2022/Day11/cpp/main.cpp
#include <iostream>
#include <fstream>
#include <set>
#include <vector>
#include <algorithm>
#include <iterator>
#include <sstream>
#include <string.h>
#include <queue>
#include <assert.h>

using namespace std;

struct Monkey
{
    queue<int> items;
    char op;
    uint8_t val_op;
    uint8_t test;
    uint8_t tt;
    uint8_t tf;
    uint32_t inspected = 0;

    void print()
    {
        cout << "op: " << op << " test: " << (int)test << " tt: " << (int)tt << " tf: " << (int)tf << " inspected: " << inspected << " items: ";
        queue<int> Q = items;
        for (size_t i = 0; i < Q.size(); i++)
        {
            cout << Q.front() << " ";
            Q.push(Q.front());
            Q.pop();
        }
        cout << endl;
    }

    void execute(vector<Monkey> *other_monkeys)
    {
        while (!items.empty())
        {

            int item = items.front();
            items.pop();
            int64_t val = 0;
            switch (op)
            {
            case '+':
                if (val_op != 0)
                    val = item + val_op;
                else
                    val = item + item;
                break;

            case '*':
                if (val_op != 0)
                    val = item * val_op;
                else
                    val = item * item;
                break;

            default:
                cout << "Error: op not valid \n";
                exit(-1);
                break;
            }

            val = val / 3;
            if (val % test == 0)
            {
                other_monkeys->at(tt).items.push(val);
            }
            else
            {
                other_monkeys->at(tf).items.push(val);
            }

            inspected++;
        }
    }
};

int solve_one(string input_file)
{
    ifstream file(input_file);
    if (file.is_open())
    {
        vector<Monkey> monkeys;
        string line, tmp;

        // with this we skip always the first line "Monkey X"
        while (getline(file, line))
        {
            if (line.size() < 2)
            {
                getline(file, line); // skip empty
            }

            Monkey m;
            std::stringstream ss(line);

            getline(file, line); // items
            std::stringstream items(line);
            items >> tmp >> tmp; // skip Starting items:
            while (items >> tmp)
            {
                char *token;
                char *buf = strdup(tmp.c_str());
                token = strtok(buf, ",");
                m.items.push(atoi(token));
            }
            getline(file, line); // operations
            std::stringstream operation(line);
            operation >> tmp >> tmp >> tmp >> tmp; // skip "Operation: new = old"
            operation >> tmp;
            m.op = tmp.c_str()[0];
            operation >> tmp;
            m.val_op = atoi(tmp.c_str());

            getline(file, line); // test
            std::stringstream test(line);
            test >> tmp >> tmp >> tmp; // skip "Test: divisible by"
            test >> tmp;
            m.test = atoi(tmp.c_str());

            getline(file, line); // tt
            std::stringstream tt(line);
            tt >> tmp >> tmp >> tmp >> tmp >> tmp; // skip "If true: throw to monkey"
            tt >> tmp;
            m.tt = atoi(tmp.c_str());

            getline(file, line); // tf
            std::stringstream tf(line);
            tf >> tmp >> tmp >> tmp >> tmp >> tmp; // skip "If false: throw to monkey"
            tf >> tmp;
            m.tf = atoi(tmp.c_str());
            monkeys.push_back(m);
        }

        for (size_t i = 0; i < 20; i++)
        {
            for (size_t i = 0; i < monkeys.size(); i++)
            {
                monkeys[i].execute(&monkeys);
            }
        }

        vector<int> inspected;
        for (auto m : monkeys)
        {
            inspected.push_back(m.inspected);
        }
        sort(inspected.begin(), inspected.end(), greater<int>());

        return inspected[0] * inspected[1];

        file.close();
    }
    else
    {
        cout << "Error opening file \n";
        return -1;
    }
}

struct Monkey2
{
    queue<int> items;
    char op;
    uint8_t val_op;
    uint8_t test;
    uint8_t tt;
    uint8_t tf;
    uint32_t inspected = 0;
    uint32_t lcm;

    void print()
    {
        cout << "op: " << op << " test: " << (int)test << " tt: " << (int)tt << " tf: " << (int)tf << " inspected: " << inspected << " items: ";
        queue<int> Q = items;
        for (size_t i = 0; i < Q.size(); i++)
        {
            cout << Q.front() << " ";
            Q.push(Q.front());
            Q.pop();
        }
        cout << endl;
    }

    void execute(vector<Monkey2> *other_monkeys)
    {
        while (!items.empty())
        {

            uint64_t item = items.front();
            items.pop();
            uint64_t val = 0;
            switch (op)
            {
            case '+':
                if (val_op != 0)
                    val = item + val_op;
                else
                    val = item + item;
                break;

            case '*':
                if (val_op != 0)
                    val = item * val_op;
                else
                    val = item * item;
                break;

            default:
                cout << "Error: op not valid \n";
                exit(-1);
                break;
            }
            val = val % lcm;
            if (val % test == 0)
            {
                other_monkeys->at(tt).items.push(val);
            }
            else
            {
                other_monkeys->at(tf).items.push(val);
            }

            inspected++;
        }
    }
};

uint64_t solve_two(string input_file)
{
    ifstream file(input_file);
    if (file.is_open())
    {
        vector<Monkey2> monkeys;
        string line, tmp;

        int lcm = 1;

        // with this we skip always the first line "Monkey X"
        while (getline(file, line))
        {
            if (line.size() < 2)
            {
                getline(file, line); // skip empty
            }

            Monkey2 m;
            std::stringstream ss(line);

            getline(file, line); // items
            std::stringstream items(line);
            items >> tmp >> tmp; // skip Starting items:
            while (items >> tmp)
            {
                char *token;
                char *buf = strdup(tmp.c_str());
                token = strtok(buf, ",");
                m.items.push(atoi(token));
            }
            getline(file, line); // operations
            std::stringstream operation(line);
            operation >> tmp >> tmp >> tmp >> tmp; // skip "Operation: new = old"
            operation >> tmp;
            m.op = tmp.c_str()[0];
            operation >> tmp;
            m.val_op = atoi(tmp.c_str());

            getline(file, line); // test
            std::stringstream test(line);
            test >> tmp >> tmp >> tmp; // skip "Test: divisible by"
            test >> tmp;
            m.test = atoi(tmp.c_str());
            lcm *= m.test;

            getline(file, line); // tt
            std::stringstream tt(line);
            tt >> tmp >> tmp >> tmp >> tmp >> tmp; // skip "If true: throw to monkey"
            tt >> tmp;
            m.tt = atoi(tmp.c_str());

            getline(file, line); // tf
            std::stringstream tf(line);
            tf >> tmp >> tmp >> tmp >> tmp >> tmp; // skip "If false: throw to monkey"
            tf >> tmp;
            m.tf = atoi(tmp.c_str());
            monkeys.push_back(m);
        }

        for (size_t i = 0; i < monkeys.size(); i++)
        {
            monkeys[i].lcm = lcm;
        }

        for (size_t i = 0; i < 10000; i++)
        {
            for (size_t i = 0; i < monkeys.size(); i++)
            {
                monkeys[i].execute(&monkeys);
            }
        }

        vector<uint64_t> inspected;
        for (auto m : monkeys)
        {
            inspected.push_back(m.inspected);
        }
        sort(inspected.begin(), inspected.end(), greater<uint64_t>());

        return inspected[0] * inspected[1];

        file.close();
    }
    else
    {
        cout << "Error opening file \n";
        return -1;
    }
}

int main(int, char **)
{
    // with ranges
    assert(solve_one("../test.txt") == 10605);
    cout << "Part 1: " << solve_one("../../input.txt") << endl;
    assert(solve_two("../test.txt") == 2713310158);
    cout << "Part 2: " << solve_two("../../input.txt") << endl;
}